cleverhans/plot/success_fail.py

```diff
@@ -0,0 +1,250 @@
+      so not a true upper bound) on the failure rates on the t < .5 portion
+      of the curve where MaxConfidence is not optimal. Matches up with
+      `fail_lower_bound`.
+    success_bounded: success rates on the non-optimal part of the curve.
+      Matches up with `fail_lower_bound` and `fail_upper_bound`.
+  """
+  success_results = report[success_name]
+  fail_name = None # pacify pylint
+  found = False
+  for fail_name in fail_names:
+    if fail_name in report:
+      found = True
+      break
+  if not found:
+    raise ValueError(fail_name + " not in report."
+                     "Available keys: " + str(report.keys()))
+  fail_results = report[fail_name]
+
+  # "good" means drawn from the distribution where we measure success rate.
+  # "bad" means drawn from the distribution where we measure failure rate.
+  # From here on out we use those terms, to avoid confusion between examples
+  # that actually failed and examples that were drawn from the distribution
+  # where we measured failure rate.
+
+  if 'all_probs' in success_results:
+    warnings.warn("The 'all_probs' key is included only to support "
+                  " old files from a private development codebase. "
+                  "Support for this key can be dropped at any time "
+                  " without warning.")
+    good_probs = success_results['all_probs']
+    bad_probs = fail_results['all_probs']
+    bad_corrects = fail_results['correctness_mask']
+    good_corrects = success_results['correctness_mask']
+  else:
+    good_probs = success_results['confidence']
+    bad_probs = fail_results['confidence']
+    good_corrects = success_results['correctness']
+    bad_corrects = fail_results['correctness']
+  good_triplets = [(prob, correct, True) for prob, correct
+                   in safe_zip(good_probs, good_corrects)]
+  bad_triplets = [(prob, correct, False) for prob, correct
+                  in safe_zip(bad_probs, bad_corrects)]
+  total_good = len(good_triplets)
+  total_bad = len(bad_triplets)
+  if total_good != 10000:
+    warnings.warn("Not using full test set? Found " + str(total_good) +
+                  " examples for measuring success rate")
+  if total_bad != 10000:
+    warnings.warn("Not using full test set for adversarial examples?")
+  all_triplets = good_triplets + bad_triplets
+  all_triplets = sorted(all_triplets, key=lambda x: -x[0])
+
+  # Start with the case for threshold t = 1.
+  # Examples are covered only if prob > t (strict inequality)
+  # So initially nothing is covered
+  good_covered_and_correct = 0
+  bad_covered_and_incorrect = 0
+
+  # Number of examples that are bad, incorrect, and covered by
+  # a t >= 0.5, or that were merely covered by a t < 0.5
+  failure_opportunities = 0
+
+  next_idx = 0
+
+  fail_optimal = []
+  success_optimal = []
+  fail_upper_bound = []
+  fail_lower_bound = []
+  success_bounded = []
+
+  bounded = False
+
+  while True:
```

Comment:
Might be more clear what's going on here to say
`while next_idx < len(all_triplets)`
even if it is redundant.